### PR TITLE
Implement command line options

### PR DESCRIPTION
**Pull request by IntelOrca**
Add basic command line interface, with uncompress and simulate commands:
```
usage: openloco [options] [path]
                uncompress [options] <path>
                simulate [options] <path> <ticks>

options:
           -o     Output path
--help     -h     Print help
--version         Print version
```

**Note:** about half of the changes are changing S5 code to use new Stream classes. This and a few other things in the command line code were required for the network implementation.

### CODE REVIEW

File: src/OpenLoco/OpenLoco.cpp [hunk 1]
```diff
@@ -375,7 +350,7 @@ namespace OpenLoco
     {
         if (isAlreadyRunning("Locomotion"))
         {
-            exitWithError(StringIds::game_init_failure, StringIds::loco_already_running);
+            // exitWithError(StringIds::game_init_failure, StringIds::loco_already_running);
```
Comment (duncanspumpkin) [inline src/OpenLoco/OpenLoco.cpp:353]: did you mean to delete this?

File: src/OpenLoco/OpenLoco.cpp [hunk 2]
```diff
@@ -1152,9 +1178,46 @@ namespace OpenLoco
...
+        registerHooks();
+
+        try
+        {
+            initialise();
+            loadFile(path);
+        }
+        catch (...)
+        {
+        }
```
Comment (duncanspumpkin) [inline src/OpenLoco/OpenLoco.cpp:1206]: can you add some basic error messge.

File: src/OpenLoco/Utility/Stream.hpp [hunk 3]
```diff
@@ -24,3 +31,254 @@ namespace OpenLoco::Utility
...
+    public:
+        SpanStream(stdx::span<CharT> data)
+            : SpanStream(std::make_unique<SpanStreamBuffer>(data))
+        {
+        }
+    };
+}
+
+namespace OpenLoco
+{
```
Comment (duncanspumpkin) [inline src/OpenLoco/Utility/Stream.hpp:67]: why have 3 namespace OpenLoco's just merge them into one.

File: src/OpenLoco/Utility/Stream.hpp [hunk 4]
```diff
@@ -24,3 +31,254 @@ namespace OpenLoco::Utility
...
+        void seek(int64_t pos)
+        {
+            setPosition(getPosition() + pos);
+        }
+
+    private:
+        [[noreturn]] static void throwInvalidOperation() { throw std::runtime_error("Invalid operation"); }
+    };
+
+    class BinaryStream : public Stream
```
Comment (duncanspumpkin) [inline src/OpenLoco/Utility/Stream.hpp:93]: Add final to all of the derived.

File: src/OpenLoco/Utility/Stream.hpp [hunk 5]
```diff
@@ -24,3 +31,254 @@ namespace OpenLoco::Utility
...
+                throw std::runtime_error("Failed to read data");
+            std::memcpy(buffer, reinterpret_cast<const void*>(reinterpret_cast<size_t>(_data) + _index), len);
+            _index += len;
+        }
+    };
+
+    class MemoryStream : public Stream
+    {
+    private:
+        std::vector<uint8_t> _data{};
```
Comment (duncanspumpkin) [inline src/OpenLoco/Utility/Stream.hpp:137]: Should the type not be `std::byte`

File: src/OpenLoco/OpenLoco.cpp (hunk 1 above)
Comment (IntelOrca) [reply src/OpenLoco/OpenLoco.cpp:353]: Yes and no, it is very annoying to develop and test multiplayer when you can't launch multiple instances. I would like to just delete this restriction.

File: src/OpenLoco/Utility/Stream.hpp [hunk 6]
```diff
@@ -24,3 +31,254 @@ namespace OpenLoco::Utility
...
+
+    class BinaryStream : public Stream
+    {
+    private:
+        const void* _data{};
+        size_t _index{};
+        size_t _len{};
+
+    public:
+        BinaryStream(const void* data, size_t len)
```
[the remark this answers is not in the document]
Comment (IntelOrca) [reply src/OpenLoco/Utility/Stream.hpp:102]: I can add another overload for span, but I think it is useful to be able to pass the two separately as well.

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**